downstairs/src/region.rs

```diff
@@ -895,32 +918,87 @@ impl Extent {
+        while req_run_start < requests.len() {
+            let first_req = requests[req_run_start];
+
+            // Starting from the first request in the potential run, we scan
+            // forward until we find a request with a block that isn't
+            // contiguous with the request before it. Since we're counting
+            // pairs, and the number of pairs is one less than the number of
+            // requests, we need to add 1 to get our actual run length.
+            let n_contiguous_requests = requests[req_run_start..]
+                .windows(2)
+                .take_while(|reqs| {
+                    reqs[0].offset.value + 1 == reqs[1].offset.value
+                })
+                .count()
+                + 1;
+
+            // Create our responses and push them into the output. While we're
+            // at it, check for overflows
+            let resp_run_start = responses.len();
+            for req in requests[req_run_start..][..n_contiguous_requests].iter()
+            {
+                let resp =
+                    ReadResponse::from_request(req, self.block_size as usize);
+                self.check_input(req.offset, &resp.data)?;
+                responses.push(resp);
+            }
 
-            /*
-             * XXX This read_exact only works because we have filled our
-             * buffer with data ahead of time.  If we want to use
-             * an uninitialized buffer, then we need a different
-             * read or type for the destination
-             */
-            inner.file.read_exact(&mut response.data)?;
+            // Finally we get to read the actual data. That's why we're here
+            inner.file.seek(SeekFrom::Start(
+                first_req.offset.value * self.block_size,
+            ))?;
+            let mut read_buffer = BytesMut::with_capacity(
+                n_contiguous_requests * self.block_size as usize,
+            );
+            read_buffer.resize(read_buffer.capacity(), 0);
+            inner.file.read_exact(&mut read_buffer)?;
 
-            response.encryption_contexts =
-                inner.get_encryption_contexts(request.offset.value)?;
+            // Query the block metadata
+            let enc_ctxts = inner.get_encryption_contexts(
+                first_req.offset.value,
+                n_contiguous_requests as u64,
+            )?;
+            let hashes = inner.get_hashes(
+                first_req.offset.value,
+                n_contiguous_requests as u64,
+            )?;
 
-            response.hashes = inner.get_hashes(request.offset.value)?;
+            // Now it's time to put everything into the responses.
+            // We use into_iter here to move values out of enc_ctxts/hashes,
+            // avoiding a clone(). For code consistency, we use iters for the
+            // response and data chunks too. These iters will be the same length
+            // (equal to n_contiguous_requests) so zipping is fine
```

Comment:
Will these iterators be the same length if the downstairs is storing unencrypted blocks?

Reply:
yes. the function to retrieve tx hashes always returns a `Vec<Vec<>>` that is precisely equal in length to the amount of blocks you requested, regardless of the contents of the sqlite database. If there aren't any encryption contexts, it'll just be a vec full of empty vecs